openshift3-examples/vertx-kubernetes-applications/vertx-kubernetes-client-application/src/main/java/io/vertx/example/kubernetes/client/ClientApplication.java

```diff
@@ -0,0 +1,52 @@
+package io.vertx.example.kubernetes.client;
+
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.Vertx;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.client.WebClient;
+import io.vertx.ext.web.codec.BodyCodec;
+import io.vertx.servicediscovery.ServiceDiscovery;
+import io.vertx.servicediscovery.types.HttpEndpoint;
+
+public class ClientApplication extends AbstractVerticle {
+
+  private WebClient client;
+
+  public static void main(String[] args) {
+    Vertx vertx = Vertx.vertx();
+    vertx.deployVerticle(ClientApplication.class.getName());
+  }
+
+  @Override
+  public void start() {
+    Router router = Router.router(vertx);
+    router.get("/").handler(this::invoke);
+
+    // Retrieve the service discovery
+    ServiceDiscovery.create(vertx, discovery ->
+      // Retrieve a web client
+      HttpEndpoint.getWebClient(discovery, svc -> svc.getName().equals("vertx-greeting"), ar -> {
```

Comment:
Do we have ways to mock this in integration tests? (outside of K8/OS)

Reply:
ServiceDiscovery also works in other environments (even bare metal). But it may behave a bit differently based on your environment and we need to work on aligning the behavior.

Reply:
Forgot to say that in your test you can just registry the endpoint by yourself in your test and so test:

* when it's there
* when it's not
* when it's there but.... misbehave